compiler/rustc_resolve/src/late.rs

```diff
@@ -4150,35 +4150,29 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
             PathResult::Indeterminate => bug!("indeterminate path result in resolve_qpath"),
         };
 
-        if path.len() > 1
-            && let Some(res) = result.full_res()
-            && let Some((&last_segment, prev_segs)) = path.split_last()
-            && prev_segs.iter().all(|seg| !seg.has_generic_args)
-            && res != Res::Err
-            && path[0].ident.name != kw::PathRoot
-            && path[0].ident.name != kw::DollarCrate
+        let segs_after_generic = match path.iter().position(|seg| seg.has_generic_args) {
+            Some(pos) => path.len() - 1 - pos,
+            None => 0,
+        };
+        if path.iter().all(|seg| !seg.ident.span.from_expansion())
+            && let Some(unqualified) =
+                path[1..].iter().rev().enumerate().skip(segs_after_generic).find_map(|(i, seg)| {
+                    let ns = if i == 0 { ns } else { TypeNS };
```

Comment:
Not sure what is the logic here, `i` here is counted from the end, and not even the very end, but the end before `segs_after_generic` last segments.
How does it correspond to the namespace?

Reply:
It was from the very end but that wasn't clear, I've flipped it around to count from the start & added an explanation